File: app/src/main/java/com/example/weatherapp/RetrofitClient.java
package com.example.weatherapp;

import okhttp3.OkHttpClient;
import okhttp3.logging.HttpLoggingInterceptor;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

/** Retrofit dengan base apa saja (tidak dipakai karena @Url full), tapi wajib ada. */
public class RetrofitClient {
    private static Retrofit retrofit;

    private static Retrofit get() {
        if (retrofit == null) {
            HttpLoggingInterceptor log = new HttpLoggingInterceptor();
            log.setLevel(HttpLoggingInterceptor.Level.BASIC);

            OkHttpClient ok = new OkHttpClient.Builder()
                    .addInterceptor(log)
                    .build();

            retrofit = new Retrofit.Builder()
                    .baseUrl("https://api.open-meteo.com/") // dummy base; pakai @Url full
                    .client(ok)
                    .addConverterFactory(GsonConverterFactory.create())
                    .build();
        }
        return retrofit;
    }

    public static WeatherService api() {
        return get().create(WeatherService.class);
    }
}

File: app/src/main/java/com/example/weatherapp/RetrofitClient.kt
package com.example.weatherapp

import com.squareup.moshi.Moshi
import okhttp3.OkHttpClient
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.moshi.MoshiConverterFactory

object RetrofitClient {
    private val logging = HttpLoggingInterceptor().apply {
        level = HttpLoggingInterceptor.Level.BASIC
    }
    private val http by lazy {
        OkHttpClient.Builder()
            .addInterceptor(logging)
            .build()
    }

    val geo: OpenMeteoService by lazy {
        Retrofit.Builder()
            .baseUrl("https://geocoding-api.open-meteo.com/")
            .addConverterFactory(MoshiConverterFactory.create(Moshi.Builder().build()))
            .client(http)
            .build()
            .create(OpenMeteoService::class.java)
    }

    val weather: OpenMeteoService by lazy {
        Retrofit.Builder()
            .baseUrl("https://api.open-meteo.com/")
            .addConverterFactory(MoshiConverterFactory.create(Moshi.Builder().build()))
            .client(http)
            .build()
            .create(OpenMeteoService::class.java)
    }
}
